### PR TITLE
fix(JSResourceLocator): Handle missing translations silently

### LINKED ISSUE #38119: [Bug]: NC27beta1 jsresourceloader cant find resource
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

After updating to first beta of NC27, getting this error `Could not find resource core/l10n/en.js to load` 

Gotten the error 14 times within the past 5 min after installing the update.

full error below:

```
{"reqId":"DdpCiZQfCO0taWcoeR5s","level":3,"time":"2023-05-07T10:22:54+00:00","remoteAddr":"192.168.8.67","user":"will","app":"jsresourceloader","method":"GET","url":"/settings/admin/logging","message":"Could not find resource core/l10n/en.js to load","userAgent":"Mozilla/5.0 (X11; CrOS x86_64 14541.0.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36","version":"27.0.0.2","data":{"app":"jsresourceloader"},"id":"64577bff5a471"}
```

### Steps to reproduce

1. Update to beta 1 of NC27
2. view error logs
3.

### Expected behavior

No error to occur

### Installation method

Community Docker image

### Nextcloud Server version

master

### Operating system

RHEL/CentOS

### PHP engine version

PHP 8.1

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": "true",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "enabledPreviewProviders": [
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\OpenDocument",
            "OC\\Preview\\Krita",
            "OC\\Preview\\Movie",
            "OC\\Preview\\Imaginary"
        ],
        "preview_imaginary_url": "nextcloud_imaginary:9000",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nextcloud.hollhouse.us"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "27.0.0.2",
        "overwrite.cli.url": "https:\/\/nextcloud.hollhouse.us",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "ssl",
        "mail_sendmailmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpauth": 1,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "default_phone_region": "US",
        "maintenance": false,
        "memories.ffmpeg_path": "\/usr\/bin\/ffmpeg",
        "memories.ffprobe_path": "\/usr\/bin\/ffprobe",
        "memories.transcoder": "\/config\/www\/nextcloud\/apps\/memories\/exiftool-bin\/go-vod-amd64",
        "memories.no_transcode": false,
        "memories.qsv": false,
        "memories.exiftool_no_local": true,
        "updater.release.channel": "beta",
        "theme": "",
        "loglevel": 2,
        "app_install_overwrite": [
            "cookbook",
            "tasks"
        ],
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - calendar: 4.4.0-alpha.2
  - cloud_federation_api: 1.10.0
  - contacts: 5.3.0-beta.1
  - contactsinteraction: 1.8.0
  - cookbook: 0.10.2
  - dav: 1.26.0
  - federatedfilesharing: 1.17.0
  - files: 1.22.0
  - files_pdfviewer: 2.8.0
  - files_sharing: 1.19.0
  - logreader: 2.12.0
  - lookup_server_connector: 1.15.0
  - mail: 3.2.0-beta.1
  - notes: 4.8.0-beta.1
  - notifications: 2.15.0
  - oauth2: 1.15.0
  - provisioning_api: 1.17.0
  - serverinfo: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - spreed: 17.0.0-beta.1
  - tasks: 0.14.5
  - text: 3.8.0
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - updatenotification: 1.17.0
  - viewer: 2.1.0
  - workflowengine: 2.9.0
Disabled:
  - activity: 2.19.0 (installed 2.17.0)
  - admin_audit: 1.17.0
  - bruteforcesettings: 2.7.0
  - circles: 27.0.0-dev (installed 25.0.0)
  - comments: 1.17.0 (installed 1.15.0)
  - dashboard: 7.7.0 (installed 7.5.0)
  - encryption: 2.15.0
  - federation: 1.17.0 (installed 1.15.0)
  - files_external: 1.19.0
  - files_rightclick: 1.6.0 (installed 1.4.0)
  - files_trashbin: 1.17.0 (installed 1.15.0)
  - files_versions: 1.20.0 (installed 1.18.0)
  - firstrunwizard: 2.16.0 (installed 2.14.0)
  - nextcloud_announcements: 1.16.0 (installed 1.14.0)
  - password_policy: 1.17.0 (installed 1.15.0)
  - photos: 2.3.0 (installed 2.0.1)
  - privacy: 1.11.0 (installed 1.9.0)
  - recommendations: 1.6.0 (installed 1.4.0)
  - related_resources: 1.2.0 (installed 1.0.4)
  - support: 1.10.0 (installed 1.8.0)
  - survey_client: 1.15.0 (installed 1.13.0)
  - suspicious_login: 5.0.0-alpha.1
  - systemtags: 1.17.0 (installed 1.15.0)
  - twofactor_totp: 9.0.0-alpha.1
  - user_ldap: 1.17.0
  - user_status: 1.7.0 (installed 1.5.0)
  - weather_status: 1.7.0 (installed 1.5.0)
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
{"reqId":"DdpCiZQfCO0taWcoeR5s","level":3,"time":"2023-05-07T10:22:54+00:00","remoteAddr":"192.168.8.67","user":"will","app":"jsresourceloader","method":"GET","url":"/settings/admin/logging","message":"Could not find resource core/l10n/en.js to load","userAgent":"Mozilla/5.0 (X11; CrOS x86_64 14541.0.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36","version":"27.0.0.2","data":{"app":"jsresourceloader"},"id":"64577bff5a471"}
```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: #38119

## Summary
Ensure that even if no app path could be loaded, missing translations are handled silently and no error is logged in that case.
I have also added a test case for the linked issue.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Template/JSResourceLocator.php [hunk 1]
```diff
@@ -99,21 +99,23 @@ public function doFind($script) {
...
+			// check combined files and fallback to plain file location
 			if (!$this->cacheAndAppendCombineJsonIfExist($app_path, $script.'.json', $app)) {
 				$this->appendScriptIfExist($app_path, $script, $app_url);
 			}
 		} catch (AppPathNotFoundException) {
-			$this->logger->error('Could not find resource {resource} to load', [
-				'resource' => $app . '/' . $script . '.js',
-				'app' => 'jsresourceloader',
-			]);
+			// pass (genral error handling happens below)
```
Comment (artonge) [inline lib/private/Template/JSResourceLocator.php]:
```suggestion
			// pass (general error handling happens below)
```

**Review by artonge — APPROVED** [review]

File: lib/private/Template/JSResourceLocator.php (hunk 1 above)
Comment (susnux) [reply lib/private/Template/JSResourceLocator.php]: Already done :wink:

File: lib/private/Template/JSResourceLocator.php [hunk 2]
```diff
@@ -99,21 +99,23 @@ public function doFind($script) {
...
-			]);
+			// pass (general error handling happens below)
 		}
+
+		// missing translations files will be ignored
+		if (strpos($script, 'l10n/') === 0) {
+			return;
+		}
+
+		$this->logger->error('Could not find resource {resource} to load', [
```
Comment (come-nc) [inline lib/private/Template/JSResourceLocator.php:115]: I do not understand what prevents this from always logging an error? Anything outside of l10n will trigger an error, no?

Comment (susnux) [reply lib/private/Template/JSResourceLocator.php:115]: Previous (NC26 and before) behavior was to throw an `ResourceNotFoundException` if that file does not exists (and is not a translation).
Should we keep this behavior? In this case the logging should be replaced with a throw.

Comment (skjnldsv) [reply lib/private/Template/JSResourceLocator.php:120]: I think your code make sense.
Now we have proper fallback and the comments makes it understandable enough

**Review by skjnldsv — APPROVED** [review]

**Comment by juliushaertl** [comment]
Seems this may introduce broken instances after upgrade, possible fix is in https://github.com/nextcloud/server/pull/38237

**Pull request merged**